Runtime/JSError.cs

```diff
@@ -0,0 +1,302 @@
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using static NodeApi.JSNativeApi;
+using static NodeApi.JSNativeApi.Interop;
+using static NodeApi.JSNativeApi.NodeApiInterop;
+
+namespace NodeApi;
+
+public enum JSErrorType { Error, TypeError, RangeError, SyntaxError }
+
+file record struct JSErrorInfo(string? Message, napi_status Status)
+{
+    public static unsafe JSErrorInfo GetLastErrorInfo()
+    {
+        napi_get_last_error_info((napi_env)JSValueScope.Current,
+                                 out napi_extended_error_info* errorInfo).ThrowIfFailed();
+        if (errorInfo == null)
+            return new JSErrorInfo(null, napi_status.napi_ok);
+
+        if (errorInfo->error_message != null)
+        {
+            string message = Encoding.UTF8.GetString(
+                MemoryMarshal.CreateReadOnlySpanFromNullTerminated(errorInfo->error_message));
+            return new JSErrorInfo(message, errorInfo->error_code);
+        }
+
+        return new JSErrorInfo(null, errorInfo->error_code);
+    }
+
+}
+
+public struct JSError
+{
+    private string? _message;
+    private readonly JSReference? _errorRef;
+
+    private const string ErrorWrapValue = "4bda9e7e-4913-4dbc-95de-891cbf66598e-errorVal";
+    private const string DefaultMessage = "Error in native callback";
+
+    public unsafe JSError(string? message = null, JSErrorType errorType = JSErrorType.Error)
+    {
+        // Fatal error instead of exception to avoid stack overflows
+        using var fataScope = new FatalIfFailedScope();
+
+        // We must retrieve the last error info before doing anything else because
+        // doing anything else will replace the last error info.
+        JSErrorInfo errorInfo = JSErrorInfo.GetLastErrorInfo();
+
+        // A pending JS exception takes precedence over any internal error status.
+        if (IsExceptionPending())
```

Comment:
`IsExceptionPending()` may throw a new exception if it fails. Either it shouldn't do that, or this code should call the lower layer API directly and handle the status result.

Reply:
Oh, now that I've reviewed as far as `FatalIfFailedScope` I guess that resolves these issues.

Reply:
Such exceptions will result in fatal error in this code. It is aligned with `node-addon-api`. They use `NAPI_FATAL_IF_FAILED` macro is such cases.

Reply:
E.g. see: https://github.com/nodejs/node-addon-api/blob/main/napi-inl.h#L2670